diff --git a/kernelSrcs/xqa/gen_cubins.py b/kernelSrcs/xqa/gen_cubins.py
index b520779..445cdb7 100755
--- a/kernelSrcs/xqa/gen_cubins.py
+++ b/kernelSrcs/xqa/gen_cubins.py
@@ -59,8 +59,6 @@ config_list = [
 
 clean_cubin = True
 
-cubin_dir = "cubin/"
-
 nvcc_bin = 'nvcc'
 nvcc_flags = '-std=c++17 -O3 -cubin -DGENERATE_CUBIN=1 -DNDEBUG --use_fast_math -Xptxas=-v --allow-unsupported-compiler --expt-relaxed-constexpr -t 0'
 # nvcc_flags = '-std=c++17 -G -cubin -DGENERATE_CUBIN=1 -Xptxas=-v --allow-unsupported-compiler --expt-relaxed-constexpr -t 0'
@@ -81,7 +79,7 @@ cpp_file_prefix_text = R"""/*
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
-namespace tensorrt_llm
+namespace xqa
 {
 namespace kernels
 {
@@ -91,10 +89,32 @@ namespace kernels
 cpp_file_suffex_text = R"""
 // clang-format on
 } // namespace kernels
-} // namespace tensorrt_llm
+} // namespace xqa
 """
 
 cubin_meta_info_struct_prefix_text = R"""
+enum Data_type
+{
+    DATA_TYPE_BOOL,
+    DATA_TYPE_FP16,
+    DATA_TYPE_FP32,
+    DATA_TYPE_INT4,
+    DATA_TYPE_INT8,
+    DATA_TYPE_INT32,
+    DATA_TYPE_BF16,
+    DATA_TYPE_E4M3,
+    DATA_TYPE_E5M2
+};
+
+constexpr int32_t kSM_80 = 80;
+constexpr int32_t kSM_86 = 86;
+constexpr int32_t kSM_87 = 87;
+constexpr int32_t kSM_89 = 89;
+constexpr int32_t kSM_90 = 90;
+constexpr int32_t kSM_101 = 101;
+constexpr int32_t kSM_120 = 120;
+constexpr int32_t kSM_121 = 121;
+
 static const struct XQAKernelMetaInfo
 {
     Data_type mDataType;
@@ -120,7 +140,8 @@ cubin_meta_info_struct_suffix_text = R"""
 is_spec_dec = False
 
 
-def generate_cubin_meta_info_line(arch: int, compile_macros: List[CompileMacro],
+def generate_cubin_meta_info_line(arch: int,
+                                  compile_macros: List[CompileMacro],
                                   function_name: str, cubin_size: int,
                                   is_last: bool, is_spec_dec: bool):
     data_type_str = None
@@ -213,7 +234,9 @@ def map_disp_value(value):
 
 
 def build_name_info(compile_macros: List[CompileMacro]):
-    short_names = [compile_macro.short_name for compile_macro in compile_macros]
+    short_names = [
+        compile_macro.short_name for compile_macro in compile_macros
+    ]
     values = []
     for compile_macro in compile_macros:
         if compile_macro.short_name == 'kvt':
@@ -381,20 +404,28 @@ def generate_compile_arch_macro_list(compile_macro_options: list):
             else:
                 input_file_name = "mha.cu"
             arch_and_macro_list.append(
-                CompileArchMacrosAndFile(arch, compile_macros, input_file_name))
+                CompileArchMacrosAndFile(arch, compile_macros,
+                                         input_file_name))
     return arch_and_macro_list
 
 
 def generate_header_file_contents(
         all_arch_macros: List[CompileArchMacrosAndFile],
-        name_size_list: List[Tuple[str, int]], is_spec_dec: bool):
+        name_size_list: List[Tuple[str, int]]):
     cubin_data_array = []
     cubin_length_array = []
     meta_line_array = []
     for i, (arch_macro,
-            name_size) in enumerate(list(zip(all_arch_macros, name_size_list))):
+            name_size) in enumerate(list(zip(all_arch_macros,
+                                             name_size_list))):
         arch = arch_macro.arch
         macros = arch_macro.macro_list
+        is_spec_dec = False
+        for macro in macros:
+            if macro.macro_name == 'SPEC_DEC':
+                is_spec_dec = bool(macro.value)
+                break
+
         #function_name = construct_name(build_func_name_prefix, arch, build_name_info(macros))
         function_name, cubin_size = name_size
         cubin_variable_name = f"{function_name}_cubin"
@@ -418,34 +449,53 @@ def generate_header_file_contents(
 
 if __name__ == "__main__":
 
+    if len(sys.argv) > 1 and sys.argv[1] == "sm12x":
+        arch_options = [120, 121]
+        cubin_dir = "../cubin_sm12x/"
+    else:
+        arch_options = [80, 86, 87, 89, 101]
+        cubin_dir = "../cubin/"
+
     if os.path.exists(cubin_dir):
         shutil.rmtree(cubin_dir)
     os.mkdir(cubin_dir)
 
-    if len(sys.argv) > 1 and sys.argv[1] == 'spec_dec':
-        is_spec_dec = True
-        nvcc_flags = '-std=c++17 -O3 -cubin -DGENERATE_CUBIN=1 -DNDEBUG -DSPEC_DEC --use_fast_math -Xptxas=-v --allow-unsupported-compiler --expt-relaxed-constexpr -t 0'
-        arch_options = [80, 86, 89, 90]
-        config_list = [[
-            CompileMacroOption('DTYPE', 'dt', ['__half', '__nv_bfloat16']),
-            CompileMacroOption('HEAD_ELEMS', 'd', [128]),
-            CompileMacroOption('BEAM_WIDTH', 'beam', [1]),
-            CompileMacroOption('CACHE_ELEM_ENUM', 'kvt', [0, 1, 2]),
-            CompileMacroOption('TOKENS_PER_PAGE', 'pagedKV',
-                               [0, 64, 128]),  # 0 denotes contiguous kv cache.
-            CompileMacroOption('HEAD_GRP_SIZE', 'nqpkv', [0]),
-            CompileMacroOption('M_TILESIZE', 'm', [16, 32]),
-        ]]
+    edgellm_config_list = [[
+        CompileMacroOption('DTYPE', 'dt', ['__half']),
+        CompileMacroOption('HEAD_ELEMS', 'd', [128, 64, 32]),
+        CompileMacroOption('BEAM_WIDTH', 'beam', [1]),
+        CompileMacroOption('CACHE_ELEM_ENUM', 'kvt', [0]),
+        CompileMacroOption('TOKENS_PER_PAGE', 'pagedKV',
+                           [0]),  # 0 denotes contiguous kv cache.
+        CompileMacroOption('HEAD_GRP_SIZE', 'nqpkv', [1, 2, 3, 4, 5, 6, 7, 8]),
+        CompileMacroOption('M_TILESIZE', 'm', [8]),
+        CompileMacroOption('SPEC_DEC', 'spec_dec', [0]),
+    ]]
+
+    edgellm_config_list_spec_dec = [[
+        CompileMacroOption('DTYPE', 'dt', ['__half']),
+        CompileMacroOption('HEAD_ELEMS', 'd', [128, 64]),
+        CompileMacroOption('BEAM_WIDTH', 'beam', [1]),
+        CompileMacroOption('CACHE_ELEM_ENUM', 'kvt', [0]),
+        CompileMacroOption('TOKENS_PER_PAGE', 'pagedKV',
+                           [0]),  # 0 denotes contiguous kv cache.
+        CompileMacroOption('HEAD_GRP_SIZE', 'nqpkv', [0]),
+        CompileMacroOption('M_TILESIZE', 'm', [32]),
+        CompileMacroOption('SPEC_DEC', 'spec_dec', [1]),
+    ]]
+
+    edgellm_config_list.extend(edgellm_config_list_spec_dec)
     arch_macro_lists = []
-    for cfg in config_list:
+    for cfg in edgellm_config_list:
         arch_macro_lists.extend(generate_compile_arch_macro_list(cfg))
     cpu_count = os.cpu_count()
     thread_count = cpu_count // 2 if cpu_count >= 2 else cpu_count
     with multiprocessing.Pool(processes=thread_count) as pool:
         name_size_list = pool.map(run_cubin_gen, arch_macro_lists)
     header_file_contents = generate_header_file_contents(
-        arch_macro_lists, name_size_list, is_spec_dec)
+        arch_macro_lists, name_size_list)
 
     with open(cubin_dir + build_func_name_prefix + '_cubin.h', "w") as f:
         f.write("".join(
-            [cpp_file_prefix_text, header_file_contents, cpp_file_suffex_text]))
+            [cpp_file_prefix_text, header_file_contents,
+             cpp_file_suffex_text]))
diff --git a/kernelSrcs/xqa/mha.cu b/kernelSrcs/xqa/mha.cu
index 69d93e9..c96b505 100644
--- a/kernelSrcs/xqa/mha.cu
+++ b/kernelSrcs/xqa/mha.cu
@@ -84,10 +84,10 @@ constexpr uint32_t cvtExpansion = exactDiv(inputElemSize, cacheElemSize);
 constexpr uint32_t preferedKHeadPartBytes = 64;
 __constant__ constexpr uint32_t cacheVTileSeqLen = 32;
 #else
-#if __CUDA_ARCH__ == 860 || __CUDA_ARCH__ == 890 || __CUDA_ARCH__ == 1200
+#if __CUDA_ARCH__ == 860 || __CUDA_ARCH__ == 890 || __CUDA_ARCH__ == 1200 || __CUDA_ARCH__ == 1210
 constexpr uint32_t preferedKHeadPartBytes = 64;
 __constant__ constexpr uint32_t cacheVTileSeqLen = 32;
-#elif __CUDA_ARCH__ == 800 || __CUDA_ARCH__ == 870 || __CUDA_ARCH__ == 900
+#elif __CUDA_ARCH__ == 800 || __CUDA_ARCH__ == 870 || __CUDA_ARCH__ == 900 || __CUDA_ARCH__ == 1010
 constexpr uint32_t preferedKHeadPartBytes = 128;
 __constant__ constexpr uint32_t cacheVTileSeqLen = 64;
 #else
diff --git a/kernelSrcs/xqa/utils.cuh b/kernelSrcs/xqa/utils.cuh
index 2ec5b40..6938be8 100644
--- a/kernelSrcs/xqa/utils.cuh
+++ b/kernelSrcs/xqa/utils.cuh
@@ -35,11 +35,11 @@ inline constexpr int32_t kBAD_PAGE_INDEX = -1;
 __constant__ constexpr float kE4M3_MAX = 448.F;
 
 #ifdef __CUDA_ARCH__
-#if __CUDA_ARCH__ == 860 || __CUDA_ARCH__ == 890 || __CUDA_ARCH__ == 1200
+#if __CUDA_ARCH__ == 860 || __CUDA_ARCH__ == 890 || __CUDA_ARCH__ == 1200 || __CUDA_ARCH__ == 1210
 constexpr uint32_t kMAX_SMEM_SIZE = (99u << 10);
 #elif __CUDA_ARCH__ == 800 || __CUDA_ARCH__ == 870
 constexpr uint32_t kMAX_SMEM_SIZE = (163u << 10);
-#elif __CUDA_ARCH__ == 900
+#elif __CUDA_ARCH__ == 900 || __CUDA_ARCH__ == 1010
 constexpr uint32_t kMAX_SMEM_SIZE = (227u << 10);
 #endif
 #endif
